lib/matplotlib/collections.py

```diff
@@ -856,36 +851,82 @@ def get_linewidth(self):
         return self._linestyles
 
-    def update_scalarmappable(self):
-        """Update colors from the scalar mappable array, if it is not None."""
+    def _set_mappable_flags(self):
+        edge0 = self._edge_is_mapped
+        face0 = self._face_is_mapped
         if self._A is None:
+            self._edge_is_mapped = False
+            self._face_is_mapped = False
+            # return False  # Nothing to map
+        else:
+            # Typical mapping: centers, not edges.
+            self._face_is_mapped = True
+            self._edge_is_mapped = False
+
+            # Make the colors None or a string. (If None, it is a default.)
+            fc = self._original_facecolor
+            if not (fc is None or isinstance(fc, str)):
+                fc = 'array'
+            ec = self._original_edgecolor
+            if not (ec is None or isinstance(ec, str)):
+                ec = 'array'
+
+            # Handle special cases.
+            if fc == 'none':
+                self._face_is_mapped = False
+                if ec in ('face', 'none', None):
+                    self._edge_is_mapped = True
+            if ec == 'face':
+                self._edge_is_mapped = self._face_is_mapped
+
+        mapped = self._face_is_mapped or self._edge_is_mapped
+        changed = (self._edge_is_mapped != edge0
+                   or self._face_is_mapped != face0)
+        return mapped or changed
+
+    def update_scalarmappable(self):
+        """
+        Update colors from the scalar mappable array, if any.
+
+        Assign colors to edges and faces based on the array and/or
+        colors that were directly set, as appropriate.
+        """
+        if not self._set_mappable_flags():
             return
-        # QuadMesh can map 2d arrays (but pcolormesh supplies 1d array)
-        if self._A.ndim > 1 and not isinstance(self, QuadMesh):
-            raise ValueError('Collections can only map rank 1 arrays')
-        if not self._check_update("array"):
-            return
-        if np.iterable(self._alpha):
-            if self._alpha.size != self._A.size:
-                raise ValueError(f'Data array shape, {self._A.shape} '
-                                 'is incompatible with alpha array shape, '
-                                 f'{self._alpha.shape}. '
-                                 'This can occur with the deprecated '
-                                 'behavior of the "flat" shading option, '
-                                 'in which a row and/or column of the data '
-                                 'array is dropped.')
-            # pcolormesh, scatter, maybe others flatten their _A
-            self._alpha = self._alpha.reshape(self._A.shape)
-
-        if self._is_filled:
-            self._facecolors = self.to_rgba(self._A, self._alpha)
-        elif self._is_stroked:
-            self._edgecolors = self.to_rgba(self._A, self._alpha)
+        # Allow possibility to call 'self.set_array(None)'.
```

Comment:
Do we specify to allow `set_array(None)`? What should the effect be? AFAICS this is not documented anywhere.

Reply:
Self._A starts out as None, in which case there is no color mapping; it is optional for a ScalarMappable.  Mapping is always initiated via a call to set_array, as far as I know.  By allowing the None argument, we can turn mapping on and off again.  Increasing this reversibility, the ability to use None to get back to a default state, is part of what I am trying to do in this PR.